tests/auth_tests/test_management.py

```diff
@@ -1528,6 +1534,174 @@ def test_permission_with_proxy_content_type_created(self):
+
+    def test_permission_rename(self):
+        ct = ContentType.objects.create(app_label="auth_tests", model="oldmodel")
+        actions = ["add", "change", "delete", "view"]
+        for action in actions:
+            Permission.objects.create(
+                codename=f"{action}_oldmodel",
+                name=f"Can {action} old model",
+                content_type=ct,
+            )
+
+        call_command("migrate", "auth_tests", verbosity=0)
+        for action in actions:
+            self.assertFalse(
+                Permission.objects.filter(codename=f"{action}_oldmodel").exists()
+            )
+            self.assertTrue(
+                Permission.objects.filter(codename=f"{action}_newmodel").exists()
+            )
+
+        call_command(
+            "migrate",
+            "auth_tests",
+            "zero",
+            database="default",
+            interactive=False,
+            verbosity=0,
+        )
+
+        for action in actions:
+            self.assertTrue(
+                Permission.objects.filter(codename=f"{action}_oldmodel").exists()
+            )
+            self.assertFalse(
+                Permission.objects.filter(codename=f"{action}_newmodel").exists()
+            )
+
+    @mock.patch(
+        "django.db.router.allow_migrate_model",
+        return_value=False,
+    )
+    def test_rename_skipped_if_router_disallows(self, _):
+        ct = ContentType.objects.create(app_label="auth_tests", model="oldmodel")
+        Permission.objects.create(
+            codename="change_oldmodel",
+            name="Can change old model",
+            content_type=ct,
+        )
+        # The rename operation should not be there when disallowed by router.
+        app_config = apps.get_app_config("auth_tests")
+        models.signals.post_migrate.disconnect(
+            self.assertOperationsInjected, sender=app_config
+        )
+
+        call_command(
+            "migrate",
+            "auth_tests",
+            database="default",
+            interactive=False,
+            verbosity=0,
+        )
+        self.assertTrue(Permission.objects.filter(codename="change_oldmodel").exists())
+        self.assertFalse(Permission.objects.filter(codename="change_newmodel").exists())
+
+        call_command(
+            "migrate",
+            "auth_tests",
+            "zero",
+            database="default",
+            interactive=False,
+            verbosity=0,
+        )
+
+    def test_rename_backward_does_nothing_if_no_permissions(self):
+        Permission.objects.filter(content_type__app_label="auth_tests").delete()
+
+        call_command(
+            "migrate",
+            "auth_tests",
+            "zero",
+            database="default",
+            interactive=False,
+            verbosity=0,
+        )
+        self.assertFalse(
+            Permission.objects.filter(
+                codename__in=["change_oldmodel", "change_newmodel"]
+            ).exists()
+        )
```

Comment:
I think this test isn't written properly because migrating to zero with no migrations applied doesn't do anything and so no `RenamePermission` operations are injected.

Reply:
You can use "Refs #27489 -- ..." since the fix hasn't made it to a stable release yet, otherwise, we would use a new ticket.